Show the Active Health Checks on OCS Dashboard

rexagod opened the pull request:
Fixes: https://issues.redhat.com/browse/RHSTOR-2028

#### Warnings
![2021-07-08-154350_440x241_scrot](https://user-images.githubusercontent.com/33557095/124905277-84a71c00-e003-11eb-8f31-1ef25993d5ff.png)

#### Errors
![2021-07-08-155646_436x237_scrot](https://user-images.githubusercontent.com/33557095/124907472-d51f7900-e005-11eb-856b-e5368235d233.png)

rexagod commented on frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:
```diff
@@ -1,22 +1,39 @@
+  alerts.map((alert) => (
+    <Split hasGutter key={alert}>
+      <SplitItem>{icon}</SplitItem>
+      <SplitItem isFilled>{t(`ceph-storage-plugin~${alert}`)}</SplitItem>
```
@afreen23 How do we "i18n" these dynamically generated strings? Should I manually add the string in the JSON? I can see some errors related to the generated string not being present in the JSON in the console.

afreen23 replied:
The texts from operator or backend are not covered for i18n.
Hence we do not need to externalize them.
```suggestion
      <SplitItem isFilled>{alert}</SplitItem>
```

rexagod replied:
Thank you for clarifying this!

rexagod commented:
/test ceph-storage-plugin

rexagod commented:
/test ceph-storage-plugin

rexagod commented:
/retest

rexagod commented:
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_console/9426/pull-ci-openshift-console-master-ceph-storage-plugin/1413090152649068544#1:build-log.txt%3A45

rexagod commented:
/retest

rexagod commented:
Stalled till link placeholders are fixed.

rexagod commented:
/hold
Holding until links are replaced by valid ones (once they are provided).

rexagod commented:
/hold
Holding until valid links are provided for the troubleshoots.

rexagod commented:
_Introduced scrolling logic in the latest fixup._

![2021-07-14-132247_439x191_scrot](https://user-images.githubusercontent.com/33557095/125584638-7fcd8f4b-2562-46fd-afdd-9619c00961e7.png)

anmolsachan commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/whitelistedcodes.ts:
```diff
@@ -0,0 +1,5 @@
+export const whitelistedCodes = new Set(['MON_DISK_LOW', 'MON_DOWN']);
```
```suggestion
export const whitelistedHealthChecks = new Set(['MON_DISK_LOW', 'MON_DOWN']);
```

anmolsachan commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/whitelistedcodes.ts:
```diff
@@ -0,0 +1,5 @@
+export const whitelistedCodes = new Set(['MON_DISK_LOW', 'MON_DOWN']);
+
+export const whitelistedCodesRef = {
```
```suggestion
export const whitelistedHealthChecksRef = {
```

rexagod commented:
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_console/9426/pull-ci-openshift-console-master-ceph-storage-plugin/1415225555942379520#1:build-log.txt%3A45

rexagod commented:
/retest

rexagod commented:
/unhold

rexagod commented:
/retest

rexagod commented:
/retest

rexagod commented:
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_console/9426/pull-ci-openshift-console-master-images/1415563160961683456#1:build-log.txt%3A70

rexagod commented:
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_console/9426/pull-ci-openshift-console-master-ceph-storage-plugin/1415567302367121408#1:build-log.txt%3A54

rexagod commented:
/retest

rexagod commented:
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_console/9426/pull-ci-openshift-console-master-ceph-storage-plugin/1415589049850138624#1:build-log.txt%3A54

rexagod commented:
/test ceph-storage-plugin

rexagod commented:
/test ceph-storage-plugin

rexagod commented:
Failing cluster creation in CI.

rexagod commented:
/test ceph-storage-plugin

rexagod commented:
/test ceph-storage-plugin

rexagod commented:
/test ceph-storage-plugin

rexagod commented:
/test ceph-storage-plugin

GowthamShanmugam commented:
/retest

bipuladh commented:
We can do it in a much simpler manner.
`HealthItem` takes Popover related props. 
You can pass `children` as the content of the Popover and `popupTitle` to it as well. That should be enough for your use case.

rexagod commented:
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_console/9426/pull-ci-openshift-console-master-e2e-gcp-console/1417156197521297408#1:build-log.txt%3A80

rexagod commented:
/test e2e-gcp-console

rexagod commented:
/cc @bipuladh

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/whitelistedHealthChecks.ts:
```diff
@@ -0,0 +1,6 @@
+export const whitelistedHealthChecks = new Set(['MON_DISK_LOW', 'MON_DOWN']);
+
+export const whitelistedHealthChecksRef = {
+  MON_DISK_LOW: 'https://www.example.com',
+  MON_DOWN: 'https://www.example.com',
```
We should add doc links here. We can ask to create a kcs doc that can be updated later.

afreen23 replied:
KCS docs are alive and not updated every release like product docs.

anmolsachan replied:
This might not be a KCS, and most probably will be a documentation link. Once I have the confirmation, I will let you know

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/whitelistedHealthChecks.ts:
```diff
@@ -0,0 +1,6 @@
+export const whitelistedHealthChecks = new Set(['MON_DISK_LOW', 'MON_DOWN']);
```
create an enum of the Health check and use those values to refer them.
```suggestion
enum HealthChecks = {MON_DISK_LOW = 'MON_DISK_LOW'}
```

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/utils.ts:
```diff
@@ -31,7 +35,24 @@ const CephHealthStatus = (status: string, t: TFunction): SubsystemHealth => {
+  const status = cephObj?.health;
+  const details = cephObj?.details;
+
+  const getTroubleshootLink = (key: string) =>
+    whitelistedHealthChecks.has(key) ? whitelistedHealthChecksRef[key] : null;
```
We dont need a funstion for this, we can create an object.
```suggestion
const TroubleshootLinks = {
[HelathCheckType.MON_DOWN]: "https://www.wxample.com"
}
```
Then access as
```
TroubleshootLink[HealthCheck.MON_DOWN]
```

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/utils.ts:
```diff
@@ -3,12 +3,14 @@ import {
 import { WatchCephResource } from '../../../../types';
 
-const CephHealthStatus = (status: string, t: TFunction): SubsystemHealth => {
+const CephHealthStatus = (status: string, alerts: alertType[], t: TFunction): SubsystemHealth => {
```
Lets call it healthChecks`, alerts does not seems right since we are not always showing alerts
```suggestion
const CephHealthStatus = (status: string, healthChecks: HealthCheck[], t: TFunction): SubsystemHealth => {
```

afreen23 replied:
Based on below comment, please separate this logic from here.

afreen23 commented on frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:
```diff
@@ -354,7 +354,13 @@ export type PrometheusActivityProps = {
   results: PrometheusResponse[];
 };
 
+export type alertType = {
```
```suggestion
export type CephHealthCheck = {
```

afreen23 replied:
I dont think we need to do the change in common components. The health check can be seperately defined and render in ceph plugin only.

afreen23 commented on frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:
```diff
@@ -354,7 +354,13 @@ export type PrometheusActivityProps = {
 };
 
+export type alertType = {
+  alertMessage: string;
+  troubleshootLink?: string;
```
```suggestion
  detail: string;
  troubleshootLink?: string;
```

afreen23 commented on frontend/packages/ceph-storage-plugin/locales/en/console-shared.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "Troubleshoot": "Troubleshoot"
+}
```
why this ?

rexagod replied:
This might have popped up in a fixup commit. Removed.

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +79,35 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
     t,
   );
 
+  const getAlerts = (): React.ReactNode => {
```
We should make it a react component

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/utils.ts:
```diff
@@ -42,7 +63,7 @@ export const getCephHealthState: ResourceHealthHandler<WatchCephResource> = ({ c
     return { state: HealthState.NOT_AVAILABLE };
   }
-  return CephHealthStatus(status, t);
+  return CephHealthStatus(status, alerts, t);
```
We already have ceph object data in manin componnet. We dont need to build health checke here. 
`  const [data, loaded, loadError] = useK8sWatchResource<K8sResourceKind[]>(cephClusterResource);`

1.I would suggest to create a new component for health check
2. Pass ceph details data to taht component
3. Calculate and prepare the results to render

bipuladh commented on frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:
```diff
@@ -58,6 +63,7 @@ const HealthItem: React.FC<HealthItemProps> = React.memo(
 export default HealthItem;
 
 type HealthItemProps = {
+  alerts?: alertType[];
```
I don't see this used anywhere.

rexagod replied:
Removed.

bipuladh commented on frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:
```diff
@@ -17,6 +18,10 @@ const HealthItem: React.FC<HealthItemProps> = React.memo(
 
     const detailMessage = details || healthStateMessage(state, t);
 
+    const renderHealthItemIcon = (
+      <HealthItemIcon state={state} dataTest={`${title}-health-item-icon`} />
+    );
+
```
I don't see the need to refactor this.

rexagod replied:
Reverted.

bipuladh commented on frontend/packages/console-plugin-sdk/src/typings/dashboards.ts:
```diff
@@ -354,7 +354,13 @@ export type PrometheusActivityProps = {
+};
+
 export type SubsystemHealth = {
+  alerts?: alertType[];
```
Why are we using this? I am confused?

rexagod replied:
https://github.com/openshift/console/pull/9426/files#diff-878da217ade003af00a546ddc59013b7674fe3f9d905813f541733719a9012b3R13

I've made this PR much simpler than before in the next commit. Pushing in a bit.

rexagod commented:
/test e2e-gcp-console

anmolsachan commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/whitelistedHealthChecks.ts:
```diff
@@ -0,0 +1,5 @@
+// TODO(@rexagod): Add KCS/doc links once available.
+
+export enum whitelistedHealthChecksRef {
+  MON_DOWN = 'https://www.example.com',
```
```suggestion
  MON_DISK_LOW= 'https://www.example.com',
```

rexagod commented:
/retest

rexagod commented:
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_console/9426/pull-ci-openshift-console-master-ceph-storage-plugin/1418252340619644928#1:build-log.txt%3A83

rexagod commented:
/test ceph-storage-plugin

vbnrh reviewed (changes requested):
Hey, on first glance of the screenshot. can you check whether we can remove the black border around the fieldlevelhelp by modifying its css? It looks really bad

cloudbehl commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/whitelistedHealthChecks.ts:
```diff
@@ -0,0 +1,5 @@
+// TODO(@rexagod): Add KCS/doc links once available.
+
+export enum whitelistedHealthChecksRef {
+  MON_DISK_LOW = 'https://www.example.com',
```
```suggestion
  MON_DISK_LOW = '#',
```

cloudbehl commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +81,72 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
+      <>
+        <hr className="stylize-hr" />
+        <div
+          className={
+            (healthChecks as CephHealthCheck[]).length > 5 ? 'healthchecks-scrollable' : null
+          }
```
let's have function to return this

cloudbehl commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +81,72 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
+                </div>
+              </SplitItem>
+              <SplitItem isFilled>
+                <strong>{healthCheck.details}</strong>
```
```suggestion
                <strong>{healthCheck?.details}</strong>
```

rexagod commented:
> Hey, on first glance of the screenshot. can you check whether we can remove the black border around the fieldlevelhelp by modifying its css? It looks really bad

@vbnrh Fixed, PTAL.

https://user-images.githubusercontent.com/33557095/126743946-f8b08e90-4ea9-4df3-bf83-fad725ab7395.mp4

rexagod commented:
https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_console/9426/pull-ci-openshift-console-master-e2e-gcp-console/1418457174811611136#1:build-log.txt%3A43

rexagod commented:
/retest

vbnrh commented:
> > Hey, on first glance of the screenshot. can you check whether we can remove the black border around the fieldlevelhelp by modifying its css? It looks really bad
> 
> @vbnrh Fixed, PTAL.
> 
>  Screencast.2021-07-23.at.11.42.-.OpenShift.Container.Storage.Overview.OKD.mp4

Pretty cool. 
1. Another check - Is it possible to remove the underline from the link ? 

2. There are two periods in the error message ( failed to get status.. timed out: exit status 1) . Make it one and add semicolons in between the different levels of msg.( Failed to fetch status; Timed out: exit status 1) 

3. Also the alert msg looks odd to me. 
`mon a, quorum a,c`  --->  lot of commas in here

can we segregate the mon and quorum into two separate lines as one item and also capitalise the first letter of each line . 
```
(WarningIcon) Mon a
                       Quorum b,c 
(WarningIcon) Mon b is low on space 
```

If that is not possible then add semicolon in between 
```
(WarningIcon) Mon a;  Quorum b,c 
(WarningIcon) Mon b is low on space 
```

4. Is there any limit in the amount of alerts displayed here ? Can the popover overflow with messages ?

rexagod commented:
> Another check - Is it possible to remove the underline from the link ?

Fixed, PTAL.

> There are two periods in the error message ( failed to get status.. timed out: exit status 1) . Make it one and add semicolons in between the different levels of msg.( Failed to fetch status; Timed out: exit status 1)
> 
> Also the alert msg looks odd to me.
> mon a, quorum a,c ---> lot of commas in here

We're basically displaying the messages as we recieve them from the `CephCluster` CR. I've adjusted the CSS to stylize this if the message gets a bit long, among other things to do as much as possible from our end to give it a nicer look.

> Is there any limit in the amount of alerts displayed here ? Can the popover overflow with messages ?

There's a [scrolling logic in place](https://github.com/openshift/console/pull/9426/files#diff-9f04d14f583f430d6b18cd93590df7c0f80ebe0b5bcb7860fae2ce41001485a3R115) for that. Also, I've used [auto width for the `Popover` till a `MaxWidth`](https://github.com/openshift/console/pull/9426/files#diff-0980755a3a675720926320d1a00ccb28b777d0556680fb40e2a999004578e492R49-R50) for wrapping the long ones.

Thanks.

GowthamShanmugam commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/healthchecks.scss:
```diff
@@ -0,0 +1,18 @@
+
+.no-border {
+  outline: none;
+}
```
```suggestion
ceph-health-check-card {
  &__troubleshoot {
    max-height: 150px;  // put you card max size
    overflow-y: auto;  // this will takecare of scroll when content cross max-height no need > 5 check
    padding-right: var(--pf-global--spacer--md);   // md or sm your can decide
  }
  &__no-border {
    outline: none;
  }
  &__stylize-hr {
    padding: var(--pf-global--spacer--md);  // sm or md or xs
    margin: 0 var(--pf-global--spacer--sm);  // this one not sure please check
    border-top: var(--pf-global--BorderWidth--sm) solid var(--pf-global--BorderColor--100);   // 100 or 200
  }
}
```

GowthamShanmugam commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/healthchecks.scss:
```diff
@@ -0,0 +1,19 @@
+.healthchecks-scrollable {
+  overflow-y: scroll;
+  max-height: 8rem;
+}
+
+.align-troubleshoot {
+  padding-right: 0.1rem;
+}
+
+.stylize-hr {
+  padding: 0.2rem;
+  margin: 0rem;
+  border-top: 0.1rem solid grey;
+}
+
+.no-decorations {
+  outline: none;
+  text-decoration: none;
+}
```
ceph-health-check-card {
  &__troubleshoot {
    max-height: 150px;  // put you card max size
    overflow-y: auto;  // this will takecare of scroll when content cross max-height no need > 5 check
    padding-right: var(--pf-global--spacer--md);   // md or sm your can decide
  }
  &__no-border {
    outline: none;
  }
  &__stylize-hr {
    padding: var(--pf-global--spacer--md);  // sm or md or xs
    margin: 0 var(--pf-global--spacer--sm);  // this one not sure please check
    border-top: var(--pf-global--BorderWidth--sm) solid var(--pf-global--BorderColor--100);   // 100 or 200
  }
}

GowthamShanmugam replied:
```suggestion
ceph-health-check-card {
  &__troubleshoot {
    max-height: 150px;  // put you card max size
    overflow-y: auto;  // this will takecare of scroll when content cross max-height no need > 5 check
    padding-right: var(--pf-global--spacer--md);   // md or sm your can decide
  }
  &__no-border {
    outline: none;
  }
  &__stylize-hr {
    padding: var(--pf-global--spacer--md);  // sm or md or xs
    margin: 0 var(--pf-global--spacer--sm);  // this one not sure please check
    border-top: var(--pf-global--BorderWidth--sm) solid var(--pf-global--BorderColor--100);   // 100 or 200
  }
}
```

GowthamShanmugam commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +81,72 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
+    return (
+      <>
+        <hr className="stylize-hr" />
+        <div className={getClassName()}>
```
```suggestion
        <div className="ceph-health-check-card__troubleshoot">
```

GowthamShanmugam commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +81,72 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
+
+  const CephHealthChecks: React.FC = () => {
+    const healthChecks = healthChecksGetterOrPresent();
+    const getClassName = () =>
+      (healthChecks as CephHealthCheck[]).length > 5 ? 'healthchecks-scrollable' : null;
```
```suggestion

```

Max width crosses then CSS will add scroll bar using overflow-y: auto;

GowthamShanmugam commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +81,72 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
+      (healthChecks as CephHealthCheck[]).length > 5 ? 'healthchecks-scrollable' : null;
+    return (
+      <>
+        <hr className="stylize-hr" />
```
```suggestion
        
```
CSS will takecare

rexagod replied:
Introducing an `<hr/>` was Hadas' suggestion in a sync yesterday. Basically the scrollbar kind of looked odd just by itself so this was suggested to differentiate between the headers and content a bit more cleanly, I think.

GowthamShanmugam commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +81,72 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
+              </SplitItem>
+              {!!healthCheck.troubleshootLink && (
+                <SplitItem>
+                  <a className="no-decorations" href={healthCheck.troubleshootLink}>
```
```suggestion
                  <a className="ceph-health-check__no-decorations" href={healthCheck.troubleshootLink}>
```

vbnrh commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -18,10 +18,17 @@ import {
+} from '@console/shared/src/components/dashboard/status-card/states';
 import { getCephHealthState, getDataResiliencyState } from './utils';
+import { CephHealthCheck } from './types';
+import { whitelistedHealthChecksRef } from './whitelistedHealthChecks';
```
Please do this for all your component files too.

vbnrh commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +81,72 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
   );
 
+  const [isHCPresent, setHCPresence] = React.useState(false);
+  const healthChecksGetterOrPresent = (checkPresence: boolean = false) => {
```
```suggestion
  const healthChecksGetterOrPresent = (isPresent: boolean = false) => {
```

vbnrh commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/types.ts:
```diff
@@ -0,0 +1,4 @@
+export type CephHealthCheck = {
```
```suggestion
type CephHealthCheck = {
```
Move it to status card if it is not being used anywhere else.

vbnrh commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/whitelistedHealthChecks.ts:
```diff
@@ -0,0 +1,5 @@
+// TODO(@rexagod): Add KCS/doc links once available.
+
+export enum whitelistedHealthChecksRef {
+  MON_DISK_LOW = '#',
+}
```
```suggestion
export const whitelistedHealthChecksRef = {
  MON_DISK_LOW:  '#',
}
```
A key value object works better than enum.
Enums should be used for declaring constant data types for compile time checks

rexagod replied:
Afreen suggested in an earlier [review](https://github.com/openshift/console/pull/9426#discussion_r673216428) to change it to enum. Should I change it back to an object?

vbnrh replied:
yes, object is best suited as it is alert name mapped to a link. Not a Data type.

rexagod replied:
Yep, that was the originally what I went with as well. Updating.

rexagod replied:
Updating.

rexagod commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -87,7 +160,11 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
                 state={cephHealthState.state}
                 details={cephHealthState.message}
-              />
+                popupTitle={isHCPresent ? t('ceph-storage-plugin~Active health checks') : null}
```
Active health checks are essentially the ceph warnings/errors that we get from the cephcluster CR that are "active" at that particular moment. I don't think there are inactive health checks because that would mean that a warning healed itself thus it won't be a warning (or a check) anymore.

rexagod commented:
Squashed and rebased.

GowthamShanmugam commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +80,75 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
+
+  React.useEffect(
+    () => {
+      setHCPresence(healthChecksGetterOrPresent(true) as boolean);
```
you can receive a health check array and decide true or false based on an array length or !! array 

returning different types from the same function is  risky

GowthamShanmugam replied:
```suggestion
      setHCPresence(!! healthChecksGetterOrPresent());
```

rexagod replied:
I actually modified that function to only run once if `isPresent` argument that's passed is truthy, otherwise we'll have to create the whole array which isn't really needed in this usecase.

GowthamShanmugam commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +80,75 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
   );
 
+  const [isHCPresent, setHCPresence] = React.useState(false);
+  const healthChecksGetterOrPresent = (isPresent: boolean = false) => {
+    const cephDetails = data?.[0]?.status?.ceph?.details;
+    const pattern = /[A-Z]+_*|error/g;
+    const healthChecks: CephHealthCheck[] = [];
+    for (const key in cephDetails) {
+      if (pattern.test(key)) {
+        if (isPresent) {
+          return true;
+        }
+        const healthCheckObject: CephHealthCheck = {
+          details: cephDetails[key].message,
+          troubleshootLink: whitelistedHealthChecksRef[key] ?? null,
+        };
+        healthChecks.push(healthCheckObject);
+      }
+    }
+    return isPresent ? false : healthChecks;
```
```suggestion
  const healthChecksGetter = () => {
    const cephDetails = data?.[0]?.status?.ceph?.details;
    const pattern = /[A-Z]+_*|error/g;
    const healthChecks: CephHealthCheck[] = [];
    for (const key in cephDetails) {
      if (pattern.test(key)) {
        const healthCheckObject: CephHealthCheck = {
          details: cephDetails[key].message,
          troubleshootLink: whitelistedHealthChecksRef[key] ?? null,
        };
        healthChecks.push(healthCheckObject);
      }
    }
    return healthChecks;
  };
  
```

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/healthchecks.scss:
```diff
@@ -0,0 +1,18 @@
+    max-height: 6rem;
+    overflow-y: auto;
+  }
+  &__troubleshoot {
+    padding-right: var(--pf-global--spacer--xs);
+  }
+  &__no-decorations {
+    outline: none;
+    text-decoration: none;
+  }
+  &__stylize-hr {
```
follow bem. These seems like modfiers. You should have element  after `__`

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +80,75 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
+    [data?.[0]?.status?.ceph?.details],
+  );
+
+  const CephHealthChecks: React.FC = () => {
+    const healthChecks = healthChecksGetterOrPresent();
```
Its not good to define components inside components.
Move this outside. Pass ceph data in `props`

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +80,75 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
   );
 
+  const [isHCPresent, setHCPresence] = React.useState(false);
+  const healthChecksGetterOrPresent = (isPresent: boolean = false) => {
```
Make this a pure function and define outside of the component.

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -87,7 +162,11 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
+                popupTitle={isHCPresent ? t('ceph-storage-plugin~Active health checks') : null}
+                isHCPresent
+              >
+                {isHCPresent ? <CephHealthChecks /> : null}
```
```suggestion
                {isHCPresent ? <CephHealthChecks cephDetails={data?.[0]?.status?.ceph?.details} /> : null}
```

afreen23 commented on frontend/packages/console-shared/src/components/dashboard/dashboard-card/DashboardCardLink.tsx:
```diff
@@ -67,6 +70,8 @@ type DashboardCardPopupLinkProps = {
   onShow?: () => void;
   onHide?: () => void;
   position?: PopoverPosition;
+  maxWidth?: string;
+  hasAutoWidth?: boolean;
```
Why this  ?

rexagod replied:
https://github.com/openshift/console/pull/9426#discussion_r675529322

afreen23 commented on frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:
```diff
@@ -39,6 +50,9 @@ const HealthItem: React.FC<HealthItemProps> = React.memo(
                 linkTitle={title}
                 popupTitle={popupTitle}
                 className="co-status-card__popup"
+                position={isHCPresent ? PopoverPosition.bottom : PopoverPosition.top}
+                maxWidth={isHCPresent ? '40rem' : '21rem'}
+                hasAutoWidth={isHCPresent}
```
Why this ?

rexagod replied:
Hadas' suggestion from a sync that happened yesterday.

afreen23 replied:
```suggestion
                hasAutoWidth={hasAutoWidth}
```

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -47,6 +54,52 @@ export const CephAlerts = withDashboardResources(
+
+  return (
+    <>
+      <div className="ceph-health-check-card__scroll ceph-health-check-card__apply-border">
```
```suggestion
      <div className="ceph-health-check-card--scroll ceph-health-check-card--border">
```

afreen23 commented on frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:
```diff
@@ -12,7 +13,17 @@ const HealthItemIcon: React.FC<HealthItemIconProps> = ({ state, dataTest }) => (
+    noIcon = false,
+    icon,
+    children,
+    isHCPresent,
```
```suggestion
    position,
```
Better to pass a `prop` component for custom `position`.

afreen23 commented on frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:
```diff
@@ -39,6 +50,9 @@ const HealthItem: React.FC<HealthItemProps> = React.memo(
                 linkTitle={title}
                 popupTitle={popupTitle}
                 className="co-status-card__popup"
+                position={isHCPresent ? PopoverPosition.bottom : PopoverPosition.top}
```
```suggestion
                position={position  ?? PopoverPosition.top}
```

afreen23 commented on frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:
```diff
@@ -39,6 +50,9 @@ const HealthItem: React.FC<HealthItemProps> = React.memo(
                 popupTitle={popupTitle}
                 className="co-status-card__popup"
+                position={isHCPresent ? PopoverPosition.bottom : PopoverPosition.top}
+                maxWidth={isHCPresent ? '40rem' : '21rem'}
```
Same here use maxWidth as prop
```suggestion
                maxWidth={maxWidth ??  21rem}
```

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +127,12 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
     t,
   );
 
+  const [isHCPresent, setHCPresence] = React.useState(false);
+  React.useEffect(() => {
+    const cephDetails = data?.[0]?.status?.ceph?.details;
+    setHCPresence(!!cephDetails);
+  }, [data]);
+
```
It can be a value. No good reason to persist as state.
```suggestion
const isHCPresent = data?.[0]?.status?.ceph?.details;
```

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/healthchecks.scss:
```diff
@@ -0,0 +1,18 @@
+.ceph-health-check-card {
+  &--scroll {
+    max-height: 6rem;
+    overflow-y: auto;
+  }
+  &__split {
+    padding-right: var(--pf-global--spacer--xs);
+  }
+  &__link {
+    outline: none;
+    text-decoration: none;
+  }
+  &--border {
+    padding: var(--pf-global--spacer--xs);
+    margin: var(--pf-global--spacer--xs);
+    border-top: var(--pf-global--BorderWidth--sm) solid var(--pf-global--BorderColor--100);
+  }
+}
```
```suggestion
  &--scroll {
    max-height: 6rem;
    overflow-y: auto;
  }
  &--border {
    padding: var(--pf-global--spacer--xs);
    margin: var(--pf-global--spacer--xs);
    border-top: var(--pf-global--BorderWidth--sm) solid var(--pf-global--BorderColor--100);
  }
    &__split {
    padding-right: var(--pf-global--spacer--xs);
  }
  &__link {
    outline: none;
    text-decoration: none;
  }
}
```

afreen23 commented on frontend/packages/console-shared/src/components/dashboard/dashboard-card/DashboardCardLink.tsx:
```diff
@@ -31,6 +31,8 @@ export const DashboardCardPopupLink: React.FC<DashboardCardPopupLinkProps> = Rea
     onShow,
     onHide,
     position = PopoverPosition.top,
+    maxWidth,
+    hasAutoWidth,
```
cc @rawagner

afreen23 commented on frontend/packages/console-shared/src/components/dashboard/dashboard-card/DashboardCardLink.tsx:
```diff
@@ -44,7 +46,8 @@ export const DashboardCardPopupLink: React.FC<DashboardCardPopupLinkProps> = Rea
         onShow={onShow}
         onHide={onHide}
-        maxWidth="21rem"
+        maxWidth={maxWidth || '21rem'}
+        hasAutoWidth={hasAutoWidth}
```
cc @rawagner

vbnrh commented:
/approve

rexagod commented:
@spadgett The `hasAutoWidth` prop prevents additional padding in certain scenarios thus giving the tooltip a cleaner look.

https://user-images.githubusercontent.com/33557095/126822842-50a39e8e-a919-4a2c-857f-dc94eb68ab48.mp4

rexagod commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -47,6 +54,46 @@ export const CephAlerts = withDashboardResources(
+  return (
+    <div>
+      {(healthChecks as CephHealthCheck[]).map((healthCheck: CephHealthCheck) => (
+        <Flex direction={{ default: 'row' }}>
```
@afreen23 I replaced `<Split>` with `<Flex>`, PTAL.

#### Without a troubleshoot link

![2021-07-24-012258_1920x1080_scrot](https://user-images.githubusercontent.com/33557095/126834888-b86658e0-40a4-4bea-ac14-ab67a4a89149.png)

#### With a troubleshoot link

![2021-07-24-011946_1920x1080_scrot](https://user-images.githubusercontent.com/33557095/126834893-b26a9e87-862d-4885-92b1-863209cb87f2.png)

bipuladh commented on frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:
```diff
@@ -12,7 +13,19 @@ const HealthItemIcon: React.FC<HealthItemIconProps> = ({ state, dataTest }) => (
+    noIcon = false,
+    icon,
+    position,
+    hasAutoWidth,
+    maxWidth,
```
```suggestion
    hasAutoWidth,
    maxWidth,
```
HealthItem is going to be a part of the Console SDK. So changes that allow users to deviate from the standard designs shouldn't be exposed from here. I would suggest we not make changes to the width.

bipuladh commented on frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:
```diff
@@ -39,6 +52,9 @@ const HealthItem: React.FC<HealthItemProps> = React.memo(
                 linkTitle={title}
                 popupTitle={popupTitle}
                 className="co-status-card__popup"
+                position={position ?? PopoverPosition.top}
```
IMO all dashboard popovers should show on the same side.

rexagod replied:
https://github.com/openshift/console/pull/9426#discussion_r675975024

afreen23 commented:
@rexagod this looks better.
Lets be consistent with rest of console and make changes we can tell that to UXD later

afreen23 commented:
/approve cancel

afreen23 commented:
> /approve cancel

Need to resolve design inconsistencies

rexagod commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +81,60 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
 
+  const healthChecksGetterOrPresent = (checkPresence: boolean = false) => {
+    const cephDetails = data?.[0]?.status?.ceph?.details;
+    const pattern = /[A-Z]+_*|error/g;
```
This handles both warnings and alerts alike so I thought it was better to have a more general approach. Also, at a time, only one is active (Warning or Error).

rexagod replied:
```
WARNING:
  <[A-Z]+_*>: <message> // a key in case of a warning (matching the first half of the regex
ERROR:
  error: <message> // a key in case of an error (always specified as "error" in the cephDetails object)
```

This regex handles both warnings and errors, in a general manner. Not sure if a specific use case for errors is needed when it can be achieved with a single regex?

rexagod commented on frontend/packages/console-shared/src/components/dashboard/dashboard-card/DashboardCardLink.tsx:
```diff
@@ -44,7 +45,7 @@ export const DashboardCardPopupLink: React.FC<DashboardCardPopupLinkProps> = Rea
         onShow={onShow}
         onHide={onHide}
-        maxWidth="21rem"
+        maxWidth={maxWidth || '21rem'}
```
In context to this suggestion, in a sync-up with Hadas on Thursday, she suggested that:

- She's fine with the `hasAutoWidth` and the `maxWidth` modifications.
- We make the popup open below as the increasing length may hide some of the nav items/heading of the page.

rexagod commented:
@afreen23 Lets be consistent with rest of console and make changes we can tell that to UXD later.

Do you have the design with you for this? I remember the text being bold in there.

afreen23 commented:
> > Lets be consistent with rest of console and make changes we can tell that to UXD later.
> 
> Do you have the design with you for this? I remember the text being bold in there.

The tile and text botha re bolds which does not make sense. Please do the change and remove any changes toe xisting props in health item

rexagod commented on frontend/packages/console-shared/src/components/dashboard/status-card/HealthItem.tsx:
```diff
@@ -39,6 +39,7 @@ const HealthItem: React.FC<HealthItemProps> = React.memo(
                 linkTitle={title}
                 popupTitle={popupTitle}
                 className="co-status-card__popup"
+                maxWidth="30rem"
```
Removed `maxWidth` prop dependancy.

rexagod commented:
/retest

rexagod commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -47,6 +54,44 @@ export const CephAlerts = withDashboardResources(
+                .icon
+            }
+          </FlexItem>
+          <FlexItem>{healthCheck?.details}</FlexItem>
```
@afreen23 Removed `<strong>`.

rexagod commented:
/cc @afreen23 

![2021-07-24-191831_1920x1080_scrot](https://user-images.githubusercontent.com/33557095/126870446-3f56f7a3-6637-4ccf-919f-a3413903e38a.png)

rexagod commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -74,6 +81,60 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
+
+  const CephHealthChecks: React.FC = () => {
+    const healthChecks = healthChecksGetterOrPresent();
+    return (
```
The popover seems a bit odd. Below is the change I made to get this.

```
diff --git a/frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx b/frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx
index 1346ca7753..8cb43ab86f 100644
--- a/frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx
+++ b/frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx
@@ -68,7 +68,7 @@ const CephHealthChecks: React.FC<CephHealthChecksProps> = ({ cephDetails, cephHe
     }
   }
 
-  return (
+  return healthChecks && (
     <div>
       {healthChecks.map((healthCheck: CephHealthCheck) => (
         <Flex direction={{ default: 'row' }}>
@@ -135,12 +135,10 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
                 details={cephHealthState.message}
                 popupTitle={isHCPresent ? t('ceph-storage-plugin~Active health checks') : null}
               >
-                {isHCPresent ? (
-                  <CephHealthChecks
-                    cephDetails={data?.[0]?.status?.ceph?.details}
-                    cephHealthState={cephHealthState}
-                  />
-                ) : null}
+                <CephHealthChecks
+                  cephDetails={data?.[0]?.status?.ceph?.details}
+                  cephHealthState={cephHealthState}
+                />
               </HealthItem>
             </GalleryItem>
             <GalleryItem>
```

rexagod commented:
/test e2e-gcp-console

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -47,6 +54,44 @@ export const CephAlerts = withDashboardResources(
+  return (
+    <div>
+      {healthChecks.map((healthCheck: CephHealthCheck) => (
+        <Flex direction={{ default: 'row' }}>
+          <FlexItem>
+            {
+              (healthStateMapping[cephHealthState.state] || healthStateMapping[HealthState.UNKNOWN])
+                .icon
+            }
+          </FlexItem>
+          <FlexItem>{healthCheck?.details}</FlexItem>
+          <FlexItem>
+            {!!healthCheck.troubleshootLink && (
+              <a className="ceph-health-check-card__link" href={healthCheck.troubleshootLink}>
+                {t('ceph-storage-plugin~Troubleshoot')}
+              </a>
+            )}
+          </FlexItem>
+        </Flex>
```
Actually this should be a component.
Then you can do - 
```
{healthChecks.map((healthCheck: CephHealthCheck) => <CephHealthCheck details={} link={}  />

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -47,6 +54,44 @@ export const CephAlerts = withDashboardResources(
+  }
+
+  return (
+    <div>
```
```suggestion

```

rexagod replied:
Getting an error when removing it.

```
ERROR in /home/rexagod/repositories/console/frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx(77,7):
TS2322: Type '({ cephDetails, cephHealthState }: PropsWithChildren<CephHealthChecksProps>) => Element[]' is not assignable to type 'FC<CephHealthChecksProps>'.
  Type 'Element[]' is missing the following properties from type 'ReactElement<any, string | ((props: any) => ReactElement<any, string | ... | (new (props: any) => Component<any, any, any>)>) | (new (props: any) => Component<any, any, any>)>': type, props, ke
```

afreen23 replied:
yes because you retirning array of elements not an react element

afreen23 replied:
Check https://github.com/openshift/console/pull/9426#pullrequestreview-714248972
It will fix that

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -73,6 +118,7 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
     [{ response: resiliencyProgress, error: resiliencyProgressError }],
     t,
   );
+  const isHCPresent = data?.[0]?.status?.ceph?.details;
```
```suggestion
  const  hasHealthChecks = data?.[0]?.status?.ceph?.details;
```

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -87,7 +133,15 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
-              />
+                popupTitle={isHCPresent ? t('ceph-storage-plugin~Active health checks') : null}
+              >
+                {isHCPresent ? (
```
```suggestion
                {hasHelathChecks ? (
```

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -47,6 +54,50 @@ export const CephAlerts = withDashboardResources(
+
+  return (
+    <>
+      {healthChecks.map((healthCheck: CephHealthCheckType) => (
+        <CephHealthCheck cephHealthState={cephHealthState} healthCheck={healthCheck} />
```
Move this to status card component

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -87,7 +139,15 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
-              />
+                popupTitle={hasHealthChecks ? t('ceph-storage-plugin~Active health checks') : null}
+              >
+                {hasHealthChecks ? (
+                  <CephHealthChecks
+                    cephDetails={data?.[0]?.status?.ceph?.details}
+                    cephHealthState={cephHealthState}
+                  />
+                ) : null}
```
```suggestion
      {healthChecks?.map((healthCheck: CephHealthCheckType) => (
        <CephHealthCheck cephHealthState={cephHealthState} healthCheck={healthCheck} />
```

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -73,6 +124,7 @@ export const StatusCard: React.FC<DashboardItemProps> = ({
     [{ response: resiliencyProgress, error: resiliencyProgressError }],
     t,
   );
+  const hasHealthChecks = data?.[0]?.status?.ceph?.details;
```
```suggestion
  const pattern = /[A-Z]+_*|error/g;
  const healthChecks: CephHealthCheckType[] = [];
  for (const key in cephDetails) {
    if (pattern.test(key)) {
      const healthCheckObject: CephHealthCheckType = {
        details: cephDetails[key].message,
        troubleshootLink: whitelistedHealthChecksRef[key] ?? null,
      };
      healthChecks.push(healthCheckObject);
    }
  }
```

afreen23 commented on frontend/packages/ceph-storage-plugin/src/components/dashboards/persistent-internal/status-card/status-card.tsx:
```diff
@@ -47,6 +54,50 @@ export const CephAlerts = withDashboardResources(
   },
 );
 
+const CephHealthCheck: React.FC<CephHealthCheckProps> = ({ cephHealthState, healthCheck }) => {
```
Dont create this component

rexagod replied:
Do you mean `CephHealthChecks`?

afreen23 replied:
Yes.

afreen23 reviewed (approved):
/lgtm

afreen23 commented:
/lgtm cancel
squash commits !

afreen23 reviewed (approved):
/lgtm

rexagod commented:
/test e2e-gcp-console
